code/player/controller/DefaultWalkController.cs

```diff
@@ -43,9 +52,75 @@ public override void Simulate()
                 SprintSpeed = (MaxSprintSpeed - DefaultSpeed) / player.MaxStamina * player.Stamina + DefaultSpeed;
             }
 
+            IsDiving = Pawn.WaterLevel.Fraction == 1f;
+
+            if (!IsDiving)
+            {
+                TimeSinceDivingStarted = 0f;
+            }
+            else if (Host.IsServer && TimeSinceDivingStarted > DrownDamageTime)
+            {
+                using (Prediction.Off())
+                {
+                    DamageInfo damageInfo = new DamageInfo
+                    {
+                        Attacker = Pawn,
+                        Flags = DamageFlags.Drown,
+                        HitboxIndex = (int) HitboxIndex.Head,
+                        Position = Position,
+                        Damage = MathF.Max(DrownDamagePerSecond * Time.Delta, 0f)
+                    };
+
+                    Pawn.TakeDamage(damageInfo);
+                }
+            }
+
+            OnPreTickMove();
+
             base.Simulate();
         }
 
+        public void OnPreTickMove()
+        {
+            _fallVelocity = Velocity.z;
+        }
+
+        public override void CategorizePosition(bool stayOnGround)
+        {
+            base.CategorizePosition(stayOnGround);
+
+            Vector3 point = Position - Vector3.Up * 2;
+
+            if (GroundEntity != null || stayOnGround)
+            {
+                point.z -= StepSize;
+            }
+
+            TraceResult pm = TraceBBox(Position, point, 4.0f);
+
+            OnPostCategorizePosition(stayOnGround, pm);
+        }
+
+        public virtual void OnPostCategorizePosition(bool stayOnGround, TraceResult trace)
+        {
+            if (Host.IsServer && trace.Hit && _fallVelocity < -FallDamageVelocity)
+            {
+                using (Prediction.Off())
+                {
+                    DamageInfo damageInfo = new DamageInfo
+                    {
+                        Attacker = Pawn,
+                        Flags = DamageFlags.Fall,
+                        HitboxIndex = (int) HitboxIndex.LeftFoot,
+                        Position = Position,
+                        Damage = (MathF.Abs(_fallVelocity) - FallDamageVelocity) * FallDamageScale
+                    };
+
+                    Pawn.TakeDamage(damageInfo);
```

Comment:
Likewise here, another case where leaving prediction on might be the best option.